syntax = "proto3";

import "google/protobuf/descriptor.proto";
import "google/protobuf/any.proto";
import "Phantom.ProtoStore/ProtoStore.proto";

package Phantom.ProtoStore;

message Header {
	uint32 Version = 1;
	uint32 LogAlignment = 2;
	repeated uint64 LogReplayExtentNumbers = 3;
	uint64 Epoch = 5;
}

message LoggedRowWrite {
	uint64 IndexNumber = 1;
	bytes Key = 2;
	bytes Value = 3;
	uint64 SequenceNumber = 4;
	uint64 CheckpointNumber = 5;
}

message LoggedAction {
	oneof LoggedActionType {
		google.protobuf.Any ArbitraryAction = 1;
    }
}

message LoggedCheckpoint {
	uint64 IndexNumber = 1;
	repeated uint64 CheckpointNumber = 2;
}

message LogRecordExtras {
	repeated LoggedAction LoggedActions = 1;
	repeated LoggedCheckpoint LoggedCheckpoints = 2;
}

message LogRecord {
	uint32 Alignment = 1;
	// Most log records are for rows, so put them right in the message.
	repeated LoggedRowWrite Rows = 2;
	// Most log records don't have extras, so put them in another message.
	LogRecordExtras Extras = 3;
}

message InternalKeyPrefixMetadata {
}

message InternalKeySuffixMetadata {
	uint64 SequenceNumber = 1 [(FieldOptions) = {
		SortOrder: Descending;
	}];
}

message InternalKey {
	bytes KeyMessage = 20;

	uint64 SequenceNumber = 30 [(FieldOptions) = {
		SortOrder: Descending;
	}];
}

message InternalValue {
	oneof Value {
		bytes ValueMessage = 20;
    }
}

message IndexesByNameKey {
	string IndexName = 1;
}

message IndexesByNameValue {
	uint64 IndexNumber = 1;
}

message IndexesByNumberKey {
	uint64 IndexNumber = 1;
}

message IndexesByNumberValue {
	IndexSchemaDescription Schema = 1;
	string IndexName = 2;
	uint64 CreateSequenceNumber = 3;
}

message NextIndexNumberKey {}
message NextIndexNumberValue {
	uint64 NextIndexNumber = 1;
}
